File: ios/BleNative/Utils/BleExtensions.swift
import Foundation
import CoreBluetooth

extension BleError {
    var asDataObject: Any {
        return [
            "errorCode": errorCode.rawValue,
            "attErrorCode": attErrorCode,
            "iosErrorCode": iosErrorCode,
            "androidErrorCode": nil,
            "reason": reason,
            "deviceID": deviceID,
            "serviceUUID": serviceUUID,
            "characteristicUUID": characteristicUUID,
            "descriptorUUID": descriptorUUID,
            "internalMessage": internalMessage
        ]
    }
    func asErrorResult() -> Result { 
        return createErrorResult(error: asDataObject)
    }
}

extension BleState {
    var asDataObject: Any {
        switch self {
        case .unknown: return "Unknown"
        case .resetting: return "Resetting"
        case .unsupported: return "Unsupported"
        case .unauthorized: return "Unauthorized"
        case .poweredOff: return "PoweredOff"
        case .poweredOn: return "PoweredOn"
        }
    }
    
    func asSuccessResult() -> Result { return createSuccessResult(data: asDataObject) } 
}

extension CBPeripheral {
    var mtu: Int {
        if #available(iOS 9.0, *) {
            return maximumWriteValueLength(for: .withoutResponse) + 3
        } else {
            return 23
        }
    }
    
    func asDataObject(centralId: Int32) -> Any {
        return [
            "id": identifier.uuidString,
            "centralId": centralId
        ]
    }
    
    func asSuccessResult(centralId: Int32) -> Result { 
        return createSuccessResult(data: asDataObject(centralId: centralId))
    }
}

extension CBService {
    func asDataObject(centralId: Int32) -> Any {
        return [
            "id": ObjectIdGenerators.services.id(for: self),
            "centralId": centralId,
            "uuid": uuid.fullUUIDString,
            "deviceID": peripheral.identifier.uuidString,
            "isPrimary": isPrimary
            ] as [String: Any]
    }
}

extension CBCharacteristic {
    var valueBase64: String? {
        return value?.base64
    }
    
    func asDataObject(centralId: Int32) -> Any {
        return [
            "id": ObjectIdGenerators.characteristics.id(for: self),
            "centralId": centralId,
            "uuid": uuid.fullUUIDString,
            "serviceID": ObjectIdGenerators.services.id(for: service),
            "serviceUUID": service.uuid.fullUUIDString,
            "deviceID": service.peripheral.identifier.uuidString,
            "isReadable": properties.contains(.read),
            "isWritableWithResponse": properties.contains(.write),
            "isWritableWithoutResponse": properties.contains(.writeWithoutResponse),
            "isNotifiable": properties.contains(.notify),
            "isIndicatable": properties.contains(.indicate),
            ] as [String: Any]
    }
    
    func asSuccessResult(centralId: Int32) -> Result {
        return createSuccessResult(
            data: asDataObject(centralId: centralId)
        )
    } 
}

extension ScannedPeripheral {
    var mtu: Int {
        return 23
    }
    
    func asDataObject(centralId: Int32) -> Any {
        var serviceData: [String: String]?
        if let advServiceData = advertisementData[CBAdvertisementDataServiceDataKey] as? [CBUUID: Data] {
            var data = [String: String]()
            for (key, value) in advServiceData {
                data[key.fullUUIDString] = value.base64
            }
            serviceData = data
        }
        
        let manufacturerData = (advertisementData[CBAdvertisementDataManufacturerDataKey] as? Data)?.base64
        
        let serviceUUIDs = (advertisementData[CBAdvertisementDataServiceUUIDsKey] as? [CBUUID])?
            .map { (uuid: CBUUID) in uuid.fullUUIDString }
        
        let overflowServiceUUIDs = (advertisementData[CBAdvertisementDataOverflowServiceUUIDsKey] as? [CBUUID])?
            .map { (uuid: CBUUID) in uuid.fullUUIDString }
        
        let solicitedServiceUUIDs = (advertisementData[CBAdvertisementDataSolicitedServiceUUIDsKey] as? [CBUUID])?
            .map { (uuid: CBUUID) in uuid.fullUUIDString }
        
        return [
            "id": peripheral.identifier.uuidString,
            "centralId": centralId,
            "name": peripheral.name as Any,
            "rssi": rssi,
            "mtu": mtu,
            
            "manufacturerData": manufacturerData as Any,
            "serviceData": serviceData as Any,
            "serviceUUIDs": serviceUUIDs as Any,
            "localName": (advertisementData[CBAdvertisementDataLocalNameKey] as Any?) ?? NSNull(),
            "txPowerLevel": (advertisementData[CBAdvertisementDataTxPowerLevelKey] as Any?) ?? NSNull(),
            "solicitedServiceUUIDs": solicitedServiceUUIDs as Any,
            "isConnectable": (advertisementData[CBAdvertisementDataIsConnectable] as Any?) ?? NSNull(),
            "overflowServiceUUIDs": overflowServiceUUIDs as Any
            ] as [String: Any]
    }
}

extension Buffer {
    func asDataObject(centralId: Int32) -> Any {
        return [
            "id": id,
            "centralId": centralId
            ] as [String: Any]
    }
    
    func asSuccessResult(centralId: Int32) -> Result {
        return createSuccessResult(data: asDataObject(centralId: centralId))
    }
}

extension Dictionary where Key == String {
    var promiseId: String? { return self[CancelOptionKeys.promiseId.rawValue] as? String }
    var timeout: Int? { return self[CancelOptionKeys.timeout.rawValue] as? Int }
    var ignoreCancelled: Bool? { return self[CancelOptionKeys.ignoreCancelled.rawValue] as? Bool }
    var bufferStrategy: BufferActionStrategy? {
        if let strategyString = self[BufferActionKeys.strategy.rawValue] as? String {
            return BufferActionStrategy(rawValue: strategyString)
        }
        return nil
    }
    var bufferPlacement: BufferActionPlacement? {
        if let placementString = self[BufferActionKeys.placement.rawValue] as? String {
            return BufferActionPlacement(rawValue: placementString)
        }
        return nil
    }
    var bufferChunkSize: Int? { return self[BufferActionKeys.chunkSize.rawValue] as? Int }
    var emitCurrentState: Bool? { return self[MonitorStateOptions.emitCurrentState.rawValue] as? Bool }
}

File: android/src/main/java/com/polidea/blenative/utils/BleExtensions.kt
package com.polidea.blenative.utils

import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.BluetoothGattService
import com.polidea.blenative.*
import com.polidea.blenative.centralmanager.Result
import com.polidea.blenative.models.AdvertisementData
import com.polidea.blenative.models.BleScanResult
import com.polidea.blenative.models.Buffer


fun BleState.asDataObject() = value
fun BleState.asSuccessResult() = ResultUtils.createSuccessResult(asDataObject())

fun Buffer.asDataObject(centralId: Int) = mapOf<String, Any>(
        Pair("id", id),
        Pair("centralId", centralId)
)

fun Buffer.asSuccessResult(centralId: Int) = ResultUtils.createSuccessResult(asDataObject(centralId))

fun BleScanResult.asDataObject(centralId: Int): Map<String, Any?> {
    val advData = AdvertisementData.parseScanResponseData(scanRecord)

    val manufacturerData = advData.manufacturerData
    val convertedManufacturerData = if (manufacturerData != null) Base64Converter.encode(manufacturerData) else null

    val serviceData = advData.serviceData
    var convertedServiceData: Map<String, String>? = null
    if (serviceData != null) {
        val map = mutableMapOf<String, String>()
        serviceData.forEach { uuid, byteArray ->
            map[UUIDConverter.fromUUID(uuid)] = Base64Converter.encode(byteArray)
        }
        convertedServiceData = map
    }

    val serviceUUIDs = advData.serviceUUIDs
    var convertedServiceUUIDs: List<String>? = null
    if (serviceUUIDs != null) {
        convertedServiceUUIDs = serviceUUIDs.map { UUIDConverter.fromUUID(it) }
    }

    val solicitedServiceUUIDs = advData.solicitedServiceUUIDs
    var convertedSolicitedServiceUUIDs: List<String>? = null
    if (solicitedServiceUUIDs != null) {
        convertedSolicitedServiceUUIDs = solicitedServiceUUIDs.map { UUIDConverter.fromUUID(it) }
    }

    return mapOf(
            Pair("id", bleDevice.address),
            Pair("centralId", centralId),
            Pair("rssi", rssi),
            Pair("mtu", Constants.MINIMUM_MTU),
            Pair("manufacturerData", convertedManufacturerData),
            Pair("serviceData", convertedServiceData),
            Pair("serviceUUIDs", convertedServiceUUIDs),
            Pair("localName", advData.localName),
            Pair("txPowerLevel", advData.txPowerLevel),
            Pair("solicitedServiceUUIDs", convertedSolicitedServiceUUIDs),
            Pair("isConnectable", null),
            Pair("overflowServiceUUIDs", null)
    )
}

fun BluetoothDevice.asDataObject(centralId: Int) = mapOf<String, Any>(
        Pair("id", address),
        Pair("centralId", centralId)
)

fun BluetoothDevice.asSuccessResult(centralId: Int): Result {
    return ResultUtils.createSuccessResult(asDataObject(centralId))
}

fun BluetoothGattService.asDataObject(centralId: Int, deviceAddress: String) = mapOf(
        Pair("id", ObjectIdGenerators.services.idForElement(this)),
        Pair("centralId", centralId),
        Pair("uuid", uuid.toString()),
        Pair("deviceID", deviceAddress),
        Pair("isPrimary", type == BluetoothGattService.SERVICE_TYPE_PRIMARY)
)

fun BluetoothGattService.asSuccessResult(centralId: Int, deviceAddress: String): Result {
    return ResultUtils.createSuccessResult(asDataObject(centralId, deviceAddress))
}

val BluetoothGattCharacteristic.valueBase64: String
    get() = Base64Converter.encode(value)

fun BluetoothGattCharacteristic.asDataObject(centralId: Int, deviceAddress: String) = mapOf(
        Pair("id", ObjectIdGenerators.characteristics.idForElement(this)),
        Pair("centralId", centralId),
        Pair("uuid", uuid.toString()),
        Pair("serviceID", ObjectIdGenerators.services.idForElement(service)),
        Pair("serviceUUID", service.uuid.toString()),
        Pair("deviceID", deviceAddress),
        Pair("isReadable", CharacteristicUtils.isReadable(this)),
        Pair("isWritableWithResponse", CharacteristicUtils.isWriteableWithResponse(this)),
        Pair("isWritableWithoutResponse", CharacteristicUtils.isWriteableWithoutResponse(this)),
        Pair("isNotifiable", CharacteristicUtils.isNotifiable(this)),
        Pair("isIndicatable", CharacteristicUtils.isIndicatable(this))
)

fun BluetoothGattCharacteristic.asSuccessResult(centralId: Int, deviceAddress: String): Result {
    return ResultUtils.createSuccessResult(asDataObject(centralId, deviceAddress))
}

val <T> Map<String, T>.bufferStrategy: BufferActionStrategy?
    get() = BufferActionStrategy.forValue(this[BufferActionKeys.STRATEGY.value] as? String)

val <T> Map<String, T>.bufferPlacement: BufferActionPlacement?
    get() = BufferActionPlacement.forValue(this[BufferActionKeys.PLACEMENT.value] as? String)

val <T> Map<String, T>.bufferChunkSize: Int?
    get() = this[BufferActionKeys.CHUNK_SIZE.value] as Int?

val <T> Map<String, T>.timeout: Int?
    get() = this[CancelOptionKeys.TIMEOUT.value] as? Int

val <T> Map<String, T>.ignoreCancelled: Boolean?
    get() = this[CancelOptionKeys.IGNORE_CANCELLED.value] as? Boolean

val <T> Map<String, T>.emitCurrentState: Boolean?
    get() = this[MonitorStateOptionKeys.EMIT_CURRENT_STATE.value] as? Boolean

val <T> Map<String, T>.autoConnect: Boolean?
    get() = this[ConnectDeviceOptionKeys.AUTO_CONNECT.value] as? Boolean

val <T> Map<String, T>.promiseId: String?
    get() = this[CancelOptionKeys.PROMISE_ID.value] as? String
